FirebaseSharedSwift/Sources/FirebaseLogger.swift

```diff
@@ -0,0 +1,183 @@
+  }
+
+  /// Returns the log object that may be used when using OSLog directly.
+  ///
+  /// ## Example Usage
+  /// ```swift
+  /// // Create a FirebaseLogger for the product.
+  /// let logger = FirebaseLogger(category: "FirebaseProduct", categoryIdentifier: "FBP")
+  ///
+  /// // Check if messages with the chosen log level (severity) should be logged.
+  /// if FirebaseLogger.isLoggableLevel(FirebaseLoggerLevel.debug) {
+  ///   os_log(
+  ///     // Get the equivalent OSLogType for the chosen log level.
+  ///     FirebaseLogger.osLogType(FirebaseLoggerLevel.debug),
+  ///     // Write to the OSLog log object for your FirebaseLogger.
+  ///     log: logger.logObject(),
+  ///     // Add the standard Firebase log message prefix with `%{public}@` since the contents of
+  ///     // dynamic strings are redacted by default.
+  ///     "%{public}@ Logged in - User name: %@, ID: %{private}ld, admin: %{BOOL}d",
+  ///     // The standard Firebase log message prefix does not contain sensitive user data but
+  ///     // dynamic string values are redacted by default; use the `%{public}@` format specifier to
+  ///     // prevent redacting.
+  ///     logger.messagePrefix(messageID: 1),
+  ///     // The user name is sensitive user data but is redacted by default; the `%@` format
+  ///     // specifier is sufficient.
+  ///     getUserName(),
+  ///     // The user ID is sensitive user data but integer values are not redacted by default;
+  ///     // redact using the `%{private}ld` format specifier.
+  ///     getUserID(),
+  ///     // The user's administrator status is not sensitive user data and is not redacted by
+  ///     // default; the `%{BOOL}d` format specifier is sufficient.
+  ///     getUserAdmin()
+  ///   )
+  /// }
+  /// ```
+  ///
+  /// - Important: When writing log messages that **do not** contain sensitive data, it is
+  ///   preferable to use the convenience logging methods ``notice(messageID:_:_:)``,
+  ///   ``info(messageID:_:_:)``, ``debug(messageID:_:_:)``, ``warning(messageID:_:_:)`` or
+  ///   ``error(messageID:_:_:)``; these methods default to `OSLogPrivacy.public`.
+  ///
+  /// - Tip: The `Logger` struct returned by ``osLogger()`` provides a nicer API for SDKs
+  ///   targeting iOS 14+ and equivalent.
+  public func logObject() -> OSLog {
+    return FIRLogOSLogObject(category)
+  }
+
+  /// Returns a Swift Logger struct that may be used when using OSLog directly.
+  @available(iOS 14.0, macOS 11.0, macCatalyst 14.0, tvOS 14.0, watchOS 7.0, *)
+  public func osLogger() -> Logger {
+    return Logger(logObject())
+  }
+
+  // MARK: - Private Helpers
+
+  private func log(level: FirebaseLoggerLevel, messageID: Int, _ message: String,
+                   _ arguments: any CVarArg...) {
+    withVaList(arguments) { va_list in
+      let code = FIRLogMessageCode(self.categoryIdentifier, messageID)
+      FIRLogBasic(level, category, code, message, va_list)
```

Comment:
```suggestion
      FIRLogBasic(level, category, code, "\(message)", va_list)
```

Reply:
Do you have any additional context for this suggestion?